synapse/events/snapshot.py

```diff
@@ -490,3 +497,47 @@ def _decode_state_dict(
         return None
 
     return immutabledict({(etype, state_key): v for etype, state_key, v in input})
+
+
+def _build_state_group_deltas(
+    unpersisted_ctx: UnpersistedEventContext,
+) -> Dict[Tuple[int, int], StateMap]:
```

Comment:
This feels like it should be a method on `UnpersistedEventContext`?

Reply:
So I put this outside `UnpersistedEventContext` because two methods use it: `batch_persist_unpersisted_contexts`, which is a method on the class `UnpersistedEventContext`, and `persist` which is an instance method on the same class, and I couldn't find a way to easily allow both methods to use `_build_state_group_deltas` other than defining it as a class-level method, which seemed more awkward than having it defined as a helper function outside the class. Let me know which sounds better to you!

Reply:
:woman_facepalming: :woman_facepalming: :woman_facepalming: 
Sorry to have missed something so obvious - I think I got tunnel vision - thank you for clarifying this.